pkg/spoke/spokecluster/lease_controller.go

```diff
@@ -0,0 +1,155 @@
+type spokeClusterLeaseController struct {
+	clusterName              string
+	hubClient                clientset.Interface
+	hubClusterLister         clusterv1listers.SpokeClusterLister
+	lastLeaseDurationSeconds int32
+	leaseUpdater             *leaseUpdater
+}
+
+// NewSpokeClusterLeaseController creates a new spoke cluster lease controller on the spoke cluster.
+func NewSpokeClusterLeaseController(
+	clusterName string,
+	hubClient clientset.Interface,
+	hubClusterInformer clusterv1informer.SpokeClusterInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &spokeClusterLeaseController{
+		clusterName:      clusterName,
+		hubClusterLister: hubClusterInformer.Lister(),
+		leaseUpdater: &leaseUpdater{
+			hubClient:   hubClient,
+			clusterName: clusterName,
+			leaseName:   fmt.Sprintf("cluter-%s-lease", clusterName),
+			recorder:    recorder,
+		},
+	}
+
+	return factory.New().
+		WithInformers(hubClusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("SpokeClusterLeaseController", recorder)
+}
+
+// sync starts a lease update routine with the spoke cluster lease duration.
+func (c *spokeClusterLeaseController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	cluster, err := c.hubClusterLister.Get(c.clusterName)
+	// unable to get spoke cluster, make sure there is no lease update routine.
+	if err != nil {
+		c.leaseUpdater.stop()
+		return fmt.Errorf("unable to get spoke cluster %q from hub: %w", c.clusterName, err)
+	}
+
+	// the spoke cluster is not accepted, make sure there is no lease update routine.
+	acceptedCondition := helpers.FindSpokeClusterCondition(cluster.Status.Conditions, clusterv1.SpokeClusterConditionHubAccepted)
+	if !helpers.IsConditionTrue(acceptedCondition) {
+		c.leaseUpdater.stop()
+		return nil
+	}
+
+	observedLeaseDurationSeconds := cluster.Spec.LeaseDurationSeconds
+	// TODO: use CRDs defaulting or mutating admission webhook to eliminate this code.
+	if observedLeaseDurationSeconds == 0 {
+		observedLeaseDurationSeconds = 60
+	}
+
+	// if lease duration is changed, start a new lease update routine.
+	if c.lastLeaseDurationSeconds != observedLeaseDurationSeconds {
+		c.lastLeaseDurationSeconds = observedLeaseDurationSeconds
+		c.leaseUpdater.stop()
+		c.leaseUpdater.start(ctx, time.Duration(c.lastLeaseDurationSeconds)*time.Second)
+	}
+
+	return nil
+}
+
+// leaseUpdater periodically updates the lease of a spoke cluster
+type leaseUpdater struct {
+	hubClient   clientset.Interface
+	clusterName string
+	leaseName   string
+	lock        sync.Mutex
+	cancel      context.CancelFunc
+	recorder    events.Recorder
+}
+
+// start a lease update routine to update the lease of a spoke cluster periodically.
+func (u *leaseUpdater) start(ctx context.Context, leaseDuration time.Duration) {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	var updateCtx context.Context
+	updateCtx, u.cancel = context.WithCancel(ctx)
+	ctrl := factory.New().
+		WithSync(u.update).
+		ResyncEvery(leaseDuration).
```

Comment:
will we able  to add jitter here to avoid simultaneous update?

Reply:
using jitter instead